File: api/TypeDBTransaction.java
/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.vaticle.typedb.client.api;

import com.vaticle.typedb.client.api.concept.ConceptManager;
import com.vaticle.typedb.client.api.logic.LogicManager;
import com.vaticle.typedb.client.api.query.QueryManager;
import com.vaticle.typedb.client.common.exception.TypeDBClientException;

import javax.annotation.CheckReturnValue;
import java.util.function.Consumer;

import static com.vaticle.typedb.client.common.exception.ErrorMessage.Internal.UNEXPECTED_NATIVE_VALUE;

public interface TypeDBTransaction extends AutoCloseable {
    @CheckReturnValue
    boolean isOpen();

    @CheckReturnValue
    Type type();

    @CheckReturnValue
    TypeDBOptions options();

    @CheckReturnValue
    ConceptManager concepts();

    @CheckReturnValue
    LogicManager logic();

    @CheckReturnValue
    QueryManager query();

    void onClose(Consumer<Throwable> function);

    void commit();

    void rollback();

    void close();

    enum Type {
        READ(0, com.vaticle.typedb.client.jni.TransactionType.Read),
        WRITE(1, com.vaticle.typedb.client.jni.TransactionType.Write);

        private final int id;
        private final boolean isWrite;
        public final com.vaticle.typedb.client.jni.TransactionType nativeObject;

        Type(int id, com.vaticle.typedb.client.jni.TransactionType nativeObject) {
            this.id = id;
            this.nativeObject = nativeObject;

            this.isWrite = nativeObject == com.vaticle.typedb.client.jni.TransactionType.Write;
        }

        public static Type of(com.vaticle.typedb.client.jni.TransactionType transactionType) {
            for (Type type : Type.values()) {
                if (type.nativeObject == transactionType) {
                    return type;
                }
            }
            throw new TypeDBClientException(UNEXPECTED_NATIVE_VALUE);
        }

        public int id() {
            return id;
        }

        public boolean isRead() {
            return !isWrite;
        }

        public boolean isWrite() {
            return isWrite;
        }
    }
}

File: csharp/connection/TypeDBTransaction.cs
/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

using System;
using System.Collections.Generic;

using Vaticle.Typedb.Driver;
using Vaticle.Typedb.Driver.Api;
using Vaticle.Typedb.Driver.Api.Database;
using Vaticle.Typedb.Driver.Common;
using Vaticle.Typedb.Driver.Common.Exception;
using DriverError = Vaticle.Typedb.Driver.Common.Exception.Error.Driver;

namespace Vaticle.Typedb.Driver.Connection
{
    public class TypeDBTransaction : NativeObjectWrapper<Pinvoke.Transaction>, ITypeDBTransaction
    {
        private readonly TransactionType _type;
        private readonly TypeDBOptions _options;

// TODO:
//        private readonly IConceptManager _conceptManager;
//        private readonly ILogicManager _logicManager;
//        private readonly IQueryManager _queryManager;

        private readonly List<TransactionOnClose> _callbacks;

        public TypeDBTransaction(TypeDBSession session, TransactionType type, TypeDBOptions options)
            : base(NewNative(session, type, options))
        {
            _type = type;
            _options = options;

//            _conceptManager = new ConceptManager(NativeObject);
//            _logicManager = new LogicManager(NativeObject);
//            _queryManager = new QueryManager(NativeObject);

            _callbacks = new List<TransactionOnClose>();
        }

        private static Pinvoke.Transaction NewNative(
            TypeDBSession session, TransactionType type, TypeDBOptions options)
        {
            try
            {
                return Pinvoke.typedb_driver.transaction_new(
                    session.NativeObject, type.NativeObject, options.NativeObject);
            }
            catch (Pinvoke.Error e)
            {
                throw new TypeDBDriverException(e);
            }
        }

        public TransactionType Type()
        {
            return _type;
        }

        public TypeDBOptions Options()
        {
            return _options;
        }

        public bool IsOpen()
        {
            if (!NativeObject.IsOwned())
            {
                return false;
            }

            return Pinvoke.typedb_driver.transaction_is_open(NativeObject);
        }
// TODO:
//        public IConceptManager Concepts()
//        {
//            return _conceptManager;
//        }
//
//        public ILogicManager Logic()
//        {
//            return _logicManager;
//        }
//
//        public IQueryManager Query()
//        {
//            return _queryManager;
//        }

        public void OnClose(Action<Exception> function)
        {
            if (!NativeObject.IsOwned())
            {
                throw new TypeDBDriverException(DriverError.TRANSACTION_CLOSED);
            }

            try
            {
                TransactionOnClose callback = new TransactionOnClose(function);
                _callbacks.Add(callback);
                Pinvoke.typedb_driver.transaction_on_close(NativeObject, callback.Released());
            }
            catch (Pinvoke.Error e)
            {
                throw new TypeDBDriverException(e);
            }
        }

        public void Commit()
        {
            if (!NativeObject.IsOwned())
            {
                throw new TypeDBDriverException(DriverError.TRANSACTION_CLOSED);
            }

            try
            {
                Pinvoke.typedb_driver.transaction_commit(NativeObject.Released());  // TODO: .Get() after implementing VoidPromises
            }
            catch (Pinvoke.Error e)
            {
                throw new TypeDBDriverException(e);
            }
        }

        public void Rollback()
        {
            if (!NativeObject.IsOwned())
            {
                throw new TypeDBDriverException(DriverError.TRANSACTION_CLOSED);
            }

            try
            {
                Pinvoke.typedb_driver.transaction_rollback(NativeObject); // TODO: .Get() after implementing VoidPromises
            }
            catch (Pinvoke.Error e)
            {
                throw new TypeDBDriverException(e);
            }
        }

        public void Close()
        {
            if (!NativeObject.IsOwned())
            {
                return;
            }

            try
            {
                Pinvoke.typedb_driver.transaction_force_close(NativeObject);
            }
            catch (Pinvoke.Error e)
            {
                throw new TypeDBDriverException(e);
            }
            finally
            {
                _callbacks.Clear();
            }
        }

        public void Dispose()
        {
        }

        private class TransactionOnClose : Pinvoke.TransactionCallbackDirector
        {
            private readonly Action<Exception> _function;

            public TransactionOnClose(Action<Exception> function)
            {
                _function = function;
            }

            public override void callback(Pinvoke.Error e)
            {
                _function(e);
            }
        }
    }
}
